Doc 1221/add a warning about the mono subscription to all events caveats for (#488)
swan-io/docs: docs/developers/using-api/webhooks.mdx

### Subscription limits {#subscription-limits}

In order to avoid overconsumption caused by too many subscriptions to the same webhook event, there's a limit on how many subscriptions you can enable per event.
Consider following [best practices for webhook consumption](#best-practices) to optimize your subscription strategy.

The following limits apply:

- **Sandbox**: 10 subscriptions per event.

> - **Live**: 5 subscriptions per event.

Reviewer: ```suggestion
- **Live**: 5 subscriptions per event
```

- 54.194.47.212
:::

## Best practices for webhook consumption {#best-practices}

Effective webhook management ensures critical events are processed promptly and improves your integration's reliability.

> ### Event subscription strategy {#best-practices-subscription}

Reviewer: Let's make these headings more verb led - at the moment the key message is getting lost in the text.

- **Processing bottlenecks**: High-volume, low-priority events can prevent timely processing of critical events that require immediate action.

### Event prioritization {#best-practices-priority}

Assess the criticality of each event type based on your integration's business logic and user experience requirements.

Consider these factors when prioritizing events:

- **Immediate impact**: Events that affect user-facing operations or require real-time responses (e.g., payment authorizations, account suspensions).
- **Business criticality**: Events that trigger essential workflows or compliance requirements.
- **Volume considerations**: High-volume events that can be processed asynchronously without immediate user impact.

For example, `Consent.Granted` might require immediate processing to complete a user action, while `AccountInvoice.Generated` can be processed with more flexibility.

### Endpoint architecture {#best-practices-endpoints}

Create separate webhook subscriptions with dedicated endpoints for different priority levels.
This architecture prevents low-priority events from blocking critical ones and isolates failures to specific event categories.

> **Benefits of distributed webhook handling:**

Reviewer: Can we have a similar structure for each best practice - currently one talks about challenges (but not benefits), while another talks about benefits (but no challenges) - so that the reader understands the benefits as a whole, we should make sure each section follows a similar thought process. e.g. 

H: Do this action (verb led heading)
B: Describe what we mean - and tell me why this matters to me
B: How do I do this

- 54.194.47.212
:::

## Best practices for webhook consumption {#best-practices}

Effective webhook management ensures critical events are processed promptly and improves your integration's reliability.

### Subscribe only to essential events {#best-practices-subscription}

Determine exactly which events your integration needs rather than subscribing to all available events.

> This keeps your webhook traffic focused and ensures critical events aren't lost in noise.

Reviewer: Can remove this sentence, since the section below already addresses the challenges.

**Challenges of subscribing to all events:**

> - **Processing bottlenecks**: High-volume, low-priority events can delay critical events that require immediate action.

Reviewer: Still not entirely clear what the challenge is - this needs to be written from the user angle. e.g. You risk delaying critical events behind a queue of low-priority noise.

> - **Event volume disparity**: `Transaction.Booked` events can significantly outnumber `Account.Created` events, potentially delaying important low-volume events.

Author: ```suggestion
- **Event volume spikes**: Frequent events like `Transaction.Booked` can bury rarer, high-value events like `Account.Created`.
```

- **Blind spots during downtime**: When a single endpoint receives all events, any downtime creates a significant backlog across all event types.

**How to choose your events:**

- **Immediate impact**: Prioritize events affecting user-facing operations (e.g., `Consent.Granted`, `AccountMembership.Suspended`).
- **Business criticality**: Include events that trigger essential workflows or compliance requirements.
- **Volume**: Consider whether high-volume events like `Transaction.Booked` can be processed asynchronously.

> Subscriptions can be created, modified, or removed at any time—start with essential events and expand as needed.

Reviewer: ```suggestion
You can create, modify, or remove subscriptions at any time. Start with essential events and expand as needed.
```
Active

### Use separate endpoints by priority {#best-practices-endpoints}

Create separate webhook subscriptions with dedicated endpoints for different priority levels.
This ensures critical events are processed immediately, regardless of overall event volume.

**Challenges of using a single endpoint:**

- **Cascading delays**: High-volume routine events block time-sensitive critical events.

> - **Difficult troubleshooting**: Issues affect all event types simultaneously, making root cause analysis harder.

Reviewer: ```suggestion
- **Difficult troubleshooting**: Issues affect all event types simultaneously, making it harder to find the source of the problem.
```
Simplify

> - **Scaling limitations**: You can't allocate more resources to critical events without over-provisioning for everything.

Reviewer: Over-provisioning is business jargon - simplify.

**How to structure your endpoints:**

- **Isolate by priority**: Route critical events (consent, account status changes) to a dedicated high-priority endpoint.
- **Scale independently**: Allocate endpoint capacity based on specific event volume and criticality.

> - **Contain failures**: Downtime or issues for one category won't affect others.

Reviewer: ```suggestion
- **Separate by function**: Use different endpoints for distinct workflows so an issue in one area doesn't block another.
```
Are we saying we want them to add failures when structuring their endpoints? This reads more like a plus point rather than a way to structure. Have suggested an option that follows the structure of the other two, which are verb-led.
